File: GFG POTD/032-February-2026/006-happiest-triplet.py
# https://www.geeksforgeeks.org/problems/happiest-triplet2921/1/




class Solution:
    # TC : O(N*log(N))
    # SC : O(1)
    def smallestDiff(self,a, b, c):
        a.sort()
        b.sort()
        c.sort()
        
        i = j = k = 0 # 3-pointers
        x = y = z = 0 # 3-elements
        
        bestDiff = float('inf')
        bestSum  = float('inf')
        
        while i < len(a) and j < len(b) and k < len(c):

            mn = min(a[i], min(b[j], c[k]))
            mx = max(a[i], max(b[j], c[k]))
            
            currDiff = mx - mn
            currSum  = a[i] + b[j] + c[k]
            
            if currDiff < bestDiff or (currDiff == bestDiff and currSum < bestSum):
                bestSum  = currSum
                bestDiff = currDiff
                x, y, z = a[i], b[j], c[k]
                
            # move the pointer of minimum element
            if mn == a[i]:
                i += 1
            elif mn == b[j]:
                j += 1
            else:
                k += 1
                
        res = sorted([x, y, z])
        return [res[2], res[1], res[0]]

File: GFG POTD/032-February-2026/006-happiest-triplet.java
// https://www.geeksforgeeks.org/problems/happiest-triplet2921/1/




import java.util.*;

class Solution {
    int[] smallestDiff(int a[], int b[], int c[]) {
        Arrays.sort(a);
        Arrays.sort(b);
        Arrays.sort(c);

        int i = 0, j = 0, k = 0;
        int bestDiff = Integer.MAX_VALUE;
        int bestSum = Integer.MAX_VALUE;

        int x = 0, y = 0, z = 0;

        while (i < a.length && j < b.length && k < c.length) {

            int mn = Math.min(a[i], Math.min(b[j], c[k]));
            int mx = Math.max(a[i], Math.max(b[j], c[k]));
            int diff = mx - mn;
            int sum = a[i] + b[j] + c[k];

            if (diff < bestDiff || (diff == bestDiff && sum < bestSum)) {
                bestDiff = diff;
                bestSum = sum;
                x = a[i];
                y = b[j];
                z = c[k];
            }

            // move pointer of minimum element
            if (mn == a[i]) i++;
            else if (mn == b[j]) j++;
            else k++;
        }

        int[] res = new int[]{x, y, z};
        Arrays.sort(res); // increasing
        // reverse to get decreasing
        return new int[]{res[2], res[1], res[0]};
    }
}